### PR TITLE
Refactoring type switch

### PR DESCRIPTION
**Pull request by naoyam**
As part of overall refactoring of the executor, I wanted to clean up the type switch pattern like: `switch (aten_scalar_type) { case at::ScalarType::Float: ....`.

I added two utility template functions: `atenTypeDispatch` and `atenTypeDispatchWithC10Complex`. Here's what `atenTypeDispatch` looks like:

```
template <typename Functor, typename... Args>
auto atenTypeDispatch(at::ScalarType type, Functor func, Args&&... args);
```

`Functor` should be a struct with `operator()(Args...)` templated with the native C++ type of `at::ScalarType type`. `atenTypeDispatch` then calls the functor with the native type.

A variant of this dispatch is `atenTypeDispatchWithC10Complex`. This may not be necessary, but currently we are using both `c10::complex` and `std::complex`, and so sometimes we may want to dispatch with `std::complex<float>` or `c10::complex<float>`. `atenTypeDispatchWithC10Complex` does the latter.

According to @zasdfgbnm, it should be safe to treat `c10::complex<float>` and `std::complex<float>` as reinterpretable to each other. I'm not sure if that means we can just replace `c10::complex<float>` with `std::complex<float>`. In this PR, I modified two switch-based dispatch with `atenTypeDispatchWithC10Complex`. It seems one of them can be fine with `atenTypeDispatch`, but for another, the question is if the following is valid:

```
at::Tensor scalar_tensor; // type = at::ScalarType::ComplexFloat
std::complex<float> val = scalar_tensor.data_ptr<std::complex<float>>()[0];
kernel<<<>>>(val);
```

And the kernel is defined as:
```
__global__ void kernel(std::complex<float> val);
```

Here, `std::complex` is the struct defined in `complex_number.cu`. 

This is effectively what's done here (https://github.com/NVIDIA/Fuser/blob/main/csrc/executor_kernel_arg.cpp#L184), where `c10::complex<float>` is used in the host side.

So, my two questions are:

1. Is this valid? `std::complex<float> val = scalar_tensor.data_ptr<std::complex<float>>()[0];`
2. Is it valid to pass a `std::complex<float>` value on the host side to a CUDA kernel as argument to a parameter of type `std::complex<float>`

Note that on the host side, the definition of `std::complex` is not the same as our `std::complex` definition in complex_numbers.cu. Are they ABI-compatible?

That actually seems to be the case per "Array-oriented access" in https://en.cppreference.com/w/cpp/numeric/complex, but I'm not completely sure.

Currently, this PR should reproduce what the TOT does faithfully. If both usages should be fine with just using `atenTypeDispatch`, I'd happy to update the PR.

### CODE REVIEW

File: csrc/executor_kernel_arg.cpp [hunk 1]
```diff
@@ -54,27 +54,19 @@ std::unique_ptr<TensorArgAbstract> getTensorArg(int nDims) {
       return std::make_unique<TensorArg<
           TensorArgCodegen<T, 4, nvfuser_index_t>,
           nvfuser_index_t>>();
-    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
```
Comment (naoyam) [inline csrc/executor_kernel_arg.cpp:57]: clang-tidy doesn't complain without this

File: csrc/executor_kernel_arg.cpp [hunk 2]
```diff
@@ -88,36 +80,20 @@ std::unique_ptr<TensorArgAbstract> getTensorArg(int nDims) {
...
-    case c10::ScalarType::ComplexDouble:
-      return getTensorArg<c10::complex<double>, INDEX_MODE>(nDims);
-    default:
-      TORCH_CHECK(
-          false,
-          "Dtype: ",
-          dtype,
-          " not currently supported in code generated kernels.");
-  }
+  return atenTypeDispatchWithC10Complex(
```
Comment (naoyam) [inline csrc/executor_kernel_arg.cpp:95]: I believe here we can just use `std::complex<float>` as we just pass pointers from host to device.

File: csrc/executor_kernel_arg.cpp [hunk 3]
```diff
@@ -176,68 +152,24 @@ KernelArgumentHolder KernelArgumentHolder::createKernelArgumentHolder(
...
-                tensor.data_ptr<int32_t>()[0]));
-        break;
-      default:
-        TORCH_CHECK(
-            false,
-            "Dtype: ",
-            tensor.scalar_type(),
-            " not currently supported in code generated kernels.");
-    }
+    arguments_.push_back(atenTypeDispatchWithC10Complex(
```
Comment (naoyam) [inline csrc/executor_kernel_arg.cpp:171]: Here, it isn't clear to me if it's safe to use `std::complex`

**Comment by zasdfgbnm** [comment]
I don't think `data_ptr` and `item` works with `std::complex`, see https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/templates/TensorMethods.cpp
where `AT_FORALL_SCALAR_TYPES_WITH_COMPLEX` is
https://github.com/pytorch/pytorch/blob/ef3ea30eed475c0dcfe915ac310f63bd274b2e06/c10/core/ScalarType.h#L29

**Comment by zasdfgbnm** [comment]
And yes, it is valid to pass a host side `std::complex` or `c10::complex` and view it as our own `std::complex` on the device side. Thanks to [array-oriented access](https://en.cppreference.com/w/cpp/numeric/complex), regardless of how `complex` class is defined, its data is always just `T[2]`.

**Review by zasdfgbnm — APPROVED** [review]

**Comment by naoyam** [comment]
Thanks for the comments. Looks like using `c10::complex` is still necessary due to the use of `data_ptr`. I'll merge this as is.

**Pull request merged**